examples/petsc/dmswarm.c

```diff
@@ -0,0 +1,412 @@
+      PetscReal       v[dim], J[dim * dim], invJ[dim * dim], detJ;
+      PetscInt       *points;
+      PetscInt        num_points_in_cell;
+
+      PetscCall(DMSwarmSortGetPointsPerCell(dm_swarm, cell, &num_points_in_cell, &points));
+      PetscCall(DMGetWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_points_true));
+      PetscCall(DMGetWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_points_ref));
+      // ---- Reference coordinates for swarm points in background mesh element
+      for (PetscInt p = 0; p < num_points_in_cell; p++) {
+        for (PetscInt d = 0; d < dim; d++) coords_points_true[p * dim + d] = coord_points[points[p] * dim + d];
+      }
+      PetscCall(DMPlexCoordinatesToReference(dm_mesh, cell, num_points_in_cell, coords_points_true, coords_points_ref));
+
+      // ---- Interpolate values from current element in background mesh to swarm points
+      PetscCall(PetscFECreateTabulation(fe, 1, num_points_in_cell, coords_points_ref, 1, &tabulation));
+      PetscCall(DMPlexComputeCellGeometryFEM(dm_mesh, cell, NULL, v, J, invJ, &detJ));
+      PetscCall(DMPlexVecGetClosure(dm_mesh, NULL, U_loc, cell, NULL, &u_cell));
+      for (PetscInt p = 0; p < num_points_in_cell; p++) {
+        PetscScalar x[dim];
+
+        PetscCall(PetscFEGetQuadrature(fe, &quadrature));
+        PetscCall(PetscFECreateCellGeometry(fe, quadrature, &fe_geometry));
+        PetscCall(PetscFEInterpolateAtPoints_Static(fe, tabulation, u_cell, &fe_geometry, p, &u_all_points_array[points[p]]));
+        PetscCall(PetscFEDestroyCellGeometry(fe, &fe_geometry));
+        for (PetscInt d = 0; d < dim; d++) x[d] = coords_points_true[p * dim + d];
+        PetscCheck(PetscAbs(u_all_points_array[points[p]] - EvalU(dim, x)) > 1E-4, comm, PETSC_ERR_USER, "Incorrect interpolated value from PETSc");
+      }
+
+      // ---- Cleanup
+      PetscCall(DMPlexVecRestoreClosure(dm_mesh, NULL, U_loc, cell, NULL, &u_cell));
+      PetscCall(DMRestoreWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_points_true));
+      PetscCall(DMRestoreWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_points_ref));
+      PetscCall(PetscTabulationDestroy(&tabulation));
+      PetscCall(PetscFree(points));
+    }
+
+    // -- Cleanup
+    PetscCall(DMSwarmRestoreField(dm_swarm, "DMSwarmPIC_coor", NULL, NULL, (void **)&coord_points));
+    PetscCall(DMSwarmRestoreField(dm_swarm, "u", NULL, NULL, (void **)&u_all_points_array));
+    PetscCall(DMSwarmSortRestoreAccess(dm_swarm));
+    PetscCall(DMRestoreLocalVector(dm_mesh, &U_loc));
+  }
+
+  // Interpolate from mesh to points via libCEED
+  {
+    IS                 is_points;
+    Vec                U_loc, ref_coords;
+    PetscInt           num_points_local;
+    const PetscInt    *all_points;
+    PetscInt           cell_start, cell_end;
+    PetscScalar       *u_all_points_array;
+    const PetscScalar *coords_points_ref;
+    CeedVector         u_evec, u_cell;
```

Comment:
Variable naming is a bit of a thick soup right now. I'm iterating as I go to balance clarity, accuracy, and brevity.